File: C++/Task2/pokling.cpp
#include <iostream>
#include <vector>
#include <tuple>
#include <cmath>
#include <random>
#include <limits>
#include <algorithm>

using namespace std;

vector<int> primes(int n);
pair<int, vector<int>> builder_test(vector<int> prime, int bit);
int test_poklin(int n, int t, vector<int> q);

int power_mod(int a, int b, int n);
int rn(int a, int b);

void print_results(const vector<int>& res, const vector<string>& res_ver_test, const vector<int>& otvegnutie);

int main() {
    int size_primes = 500;
    vector<int> prime = primes(size_primes);

    int bit = 0;
    cin >> bit;

    vector<int> q;
    int n;
    int k = 0;
    int probability;

    vector<int> res;
    vector<string> res_ver_test;
    vector<int> otvegnutie;



    while(res.size() != 10) {
        tie(n, q) = builder_test(prime, bit);
        probability = test_poklin(n, 10, q);

        if(probability == 1) {
            if (find(res.begin(), res.end(), n) == res.end()) {
                res.push_back(n);

                probability = test_poklin(n, 1, q);
                if(probability == 1) {
                    res_ver_test.push_back("+");
                }

                else{
                    res_ver_test.push_back("-");
                }

                otvegnutie.push_back(k);
                k = 0;
            }
        }

        else{
            k++;
        }
    }

    print_results(res, res_ver_test, otvegnutie);
}

vector<int> primes(int n) {
    vector<bool> is_prime(n + 1, true);
    vector<int> primes;

    for (int p = 2; p * p <= n; ++p) {
        if (is_prime[p]) {
            for (int i = p * p; i <= n; i += p)
                is_prime[i] = false;
        }
    }

    for (int p = 2; p <= n; ++p) {
        if (is_prime[p])
            primes.push_back(p);
    }

    return primes;
}

pair<int, vector<int>> builder_test(vector<int> prime, int bit) {
    int max_index = 0;
    int max_pow = 1;
// число ф должно содержать 6 бит
    for (; (prime[max_index] < pow(2, (bit / 2) + 1)) && (max_index < prime.size()); max_index++);
    for (; pow(2, max_pow) < pow(2, (bit / 2) + 1); max_pow++);
    
    int f = 1;
    vector<int> q;

    while(true){
        int num = rn(0, max_index);
        int power = rn(1, max_pow);
        
        if (pow(prime[num], power)) {
            if(f * pow(prime[num], power) < INT32_MAX) {
                f *= pow(prime[num], power);
                q.push_back(prime[num]);
            }
        }

        if(f > pow(2, (bit / 2))){

            if(f >= pow(2, (bit / 2) + 1)){
                f = 1;
                q.clear();
            }
            
            else{
                break;
            }
        }
    }

    int R;

    do
    {
        R = rn(pow(2, (bit / 2) - 1) + 1, pow(2, bit / 2));
    } while (R % 2 != 0);

    int n = R * f + 1;
    return make_pair(n, q);
}

int test_poklin(int n, int t, vector<int> q) {
    vector<int> a;
    int aj;

    while (a.size() != t) {
        aj = rn(2, n - 1);

        if (find(a.begin(), a.end(), aj) == a.end()) {
            a.push_back(aj);
        }
    }

    for (int aj : a) {
        if (power_mod(aj, n - 1, n) != 1) {
            return 0;
            break;
        }
    }

    bool flag = true;
    int i = 0;
    for (int aj : a) {
        if (q[i] != 0 && power_mod(aj, (n - 1) / q[i], n) == 1) {
            flag = false;

            return 0;
            break;
        }
    }

    if (flag) {
        return 1;
    }

    return 1;
}

int power_mod(int a, int b, int n) {
    long long result = 1;
    while (b > 0) {
        if (b % 2 == 1)
            result = (result * a) % n;
        a = (a * a) % n;
        b /= 2;
    }
    return result;
}

int rn(int a, int b) {
    mt19937 mt_rand(random_device{}());
    return uniform_int_distribution<int>(a, b)(mt_rand);
}

void print_results(const vector<int>& res, const vector<string>& res_ver_test, const vector<int>& otvegnutie) {
    cout << "Prime Numbers\tTest Results\tOccurrences" << endl;
    cout << "----------------------------------------------" << endl;

    for (size_t i = 0; i < res.size(); ++i) {
        cout << res[i] << "\t\t" << res_ver_test[i] << "\t\t" << otvegnutie[i] << endl;
    }
}

File: rust/Task2/pokling.rs
use rand::Rng;
use std::collections::HashSet;
use std::time::SystemTime;

fn primes(n: usize) -> Vec<usize> {
    let mut is_prime = vec![true; n + 1];
    is_prime[0] = false;
    is_prime[1] = false;
    let mut primes = Vec::new();

    for p in 2..=(n as f64).sqrt() as usize {
        if is_prime[p] {
            for i in (p * p..=n).step_by(p) {
                is_prime[i] = false;
            }
        }
    }

    for p in 2..=n {
        if is_prime[p] {
            primes.push(p);
        }
    }

    primes
}

fn builder_test(prime: &Vec<usize>, bit: usize) -> (usize, Vec<usize>) {
    let mut rng = rand::thread_rng();
    let max_index = prime.iter().position(|&p| p >= 2_usize.pow((bit / 2 + 1) as u32)).unwrap();
    let max_pow = (2_usize.pow((bit / 2 + 1) as u32)).leading_zeros() as usize;
    let mut f = 1;
    let mut q = Vec::new();

    loop {
        let num = rng.gen_range(0..max_index);
        let power = rng.gen_range(1..=max_pow);
        let p = prime[num].pow(power as u32);

        if f * p < usize::MAX {
            f *= p;
            q.push(prime[num]);
        }

        if f > 2_usize.pow((bit / 2) as u32) {
            if f >= 2_usize.pow((bit / 2 + 1) as u32) {
                f = 1;
                q.clear();
            } else {
                break;
            }
        }
    }

    let r = rng.gen_range(2_usize.pow((bit / 2 - 1) as u32) + 1..2_usize.pow((bit / 2) as u32));
    let n = r * f + 1;
    (n, q)
}

fn test_poklin(n: usize, t: usize, q: &[usize]) -> usize {
    let mut a = Vec::new();

    while a.len() != t {
        let aj = rng(2, n - 1);
        if !a.contains(&aj) {
            a.push(aj);
        }
    }

    for aj in &a {
        if power_mod(*aj, n - 1, n) != 1 {
            return 0;
        }
    }

    let mut flag = true;
    let mut i = 0;
    for aj in &a {
        if q.get(i).unwrap_or(&0) != &0 && power_mod(*aj, (n - 1) / q[i], n) == 1 {
            flag = false;
            return 0;
        }
    }

    if flag {
        1
    } else {
        1
    }
}

fn power_mod(a: usize, b: usize, n: usize) -> usize {
    let mut result = 1;
    let mut a = a;
    let mut b = b;

    while b > 0 {
        if b % 2 == 1 {
            result = (result * a) % n;
        }
        a = (a * a) % n;
        b /= 2;
    }

    result
}

fn rng(a: usize, b: usize) -> usize {
    let mut rng = rand::thread_rng();
    rng.gen_range(a..=b)
}

fn main() {
    let prime = primes(500);
    let bit = read_input();
    let mut res: HashSet<usize> = HashSet::new();
    let mut res_ver_test = Vec::new();
    let mut otvegnutie = Vec::new();
    let mut k = 0;

    while res.len() < 10 {
        let (n, q) = builder_test(&prime, bit);
        let probability = test_poklin(n, 10, &q);

        if probability == 1 {
            if !res.contains(&n) {
                res.insert(n);
                let probability_verify = test_poklin(n, 1, &q);
                if probability_verify == 1 {
                    res_ver_test.push("+".to_string());
                } else {
                    res_ver_test.push("-".to_string());
                }
                otvegnutie.push(k);
                k = 0;
            }
        } else {
            k += 1;
        }
    }

    print_results(&res.into_iter().collect::<Vec<_>>(), &res_ver_test, &otvegnutie);
}

fn read_input() -> usize {
    let mut input = String::new();
    std::io::stdin()
        .read_line(&mut input)
        .expect("Failed to read input");
    input.trim().parse().expect("Invalid input")
}

fn print_results(res: &[usize], res_ver_test: &[String], otvegnutie: &[usize]) {
    println!("Prime Numbers\tTest Results\tOccurrences");
    println!("----------------------------------------------");

    for i in 0..res.len() {
        println!("{}\t\t{}\t\t{}", res[i], res_ver_test[i], otvegnutie[i]);
    }
}
